### ISSUE TITLE
Set up input datasets

### ISSUE THREAD

**Issue by cscherrer** [GaloisInc/ppaml-cp6#6]
We'll need data set up to work with the evaluation framework Philip has set up. We'll need three instances of the same directory structure:

* _demo_ should have only very small files, and should be included in the Github repository.
* _public_ should have training data and the test data for team use. This should be on Midas, once it's confirmed the file formats work with a `run.sh` that calls the CRF baseline implementation.
* _private_ should have the same training data, but a different version of test data to be used for evaluation. This should also be on Midas, but should have permissions set so it is only visible to TA1 teams. 

For file names, let's make them identical, except for a prefix indicating which data set they belong to.

As an example, suppose we have something like a linear regression, where we might have `.x` files for features, and `.y` files for responses. Then the tree could look something like this:

```
public
+--run_in
|  +--train
|  |  +--public-train.x
|  |  +--public-train.y
|  |
|  +--test
|     +--public-test.x
|
+--eval_in
   +--public-test.y 
```

We went through a few iterations of this. I think we've now converged, but please let me know if any part of this is unclear or just sounds like a bad idea. Thanks!

**Comment by collinsr** [comment]
So here's how I'm mapping this to the files we've released for CP6:

* *demo* would be a small subset of cp6_*_round_1_train
* *public* (split into *public-train* and *public-test*) corresponds to all of cp6_*_round_1_train
* *private-test* corresponds to all of cp6_*_round_1_test

...but this doesn't quite work out, because public-train is only a subset of cp6_*_round_1_train. In other words, whatever data is used in public-test can't be used for training in the evaluation against the sequestered data, because public-train is fixed for both public and private... Here's an example with round numbers and ignoring the two phases of the challenge problem.

Let's say we have 15,000 images for CP6. We released 7000 of them already on midas (cp6_train.) We've reserved 8000 of them for sequestered evaluation (cp6_test.) My understanding of the above is:

* we'd put (say) 100 of the 7000 as "demo" out on git
* we'd put all 7000 of them out as "public"
  *  this 7000 would be split somehow into (say) 5000 as public-train and 2000 as public-test
* we'd copy the 5000 from public-train into private-train,  and use the sequestered 8000 as private-test.

This means that the 2000 images in public-test can't be used for training against the 8000 sequestered images.  Is that the intention?

I can see the arguments in favor: the model submitted for testing against the sequestered data has known performance against public-test, whereas if they submitted a model trained against all 7000 public images, they don't have any way to evaluate the performance of the larger model.

My concern is that we already split the data once for round 1 / round 2; I'm not sure how we'd split the released round 1 data into train / test. We could check with Tom.

(In an ideal world, we'd have a web service where they could submit models trained on any amount of the 7000 images to be evaluated against random subsets of the sequestered 8000 images, the way ImageNet works... of course, you'd have to guard against cheating, e.g. http://www.nytimes.com/2015/06/04/technology/computer-scientists-are-astir-after-baidu-team-is-barred-from-ai-competition.html ...)

**Comment by cscherrer** [comment]
Good point. How about splitting 5000/2000, and then using all 7000 of those as training in the evaluation set?

@tdietterich any thoughts on this?

**Comment by cscherrer** [comment]
@probinso has pointed out that while "real data" would be fine for the *demo* data on Github, the critical thing is just to have something to test any parsing that needs to be done. Consistency with the format is much more important than consistency with data values. And it really doesn't have to be big at all - ten data points would be completely reasonable.

**Comment by collinsr** [comment]
# Image data partitioning #

"CP6 images": the complete set of 12,889 images for CP6.

"round 1 / round 2": The CP6 challenge problem instances for
evaluation in January and July 2016, respectively.

"train / test": Training data is released to the teams. Test data is
sequested to Galois / Kitware.

"round-1-train": A set of 5216 images.
"round-1-test":  A set of 2252 images.

round-1-test / round-1-train are disjoint.

"round-2-train": A set of 9182 images, including all of round-1-train.
"round-2-test":  A set of 3707 images, including all of round-1-test.

round-2-test / round-2-train are disjoint. Their union is the
complete set of 12,889 CP6 images.


# Evaluation data file layout #

These files need to be apportioned into six separate "CP6-sandbox" structures:

round-1-demo
round-1-public
round-1-private
round-2-demo
round-2-public
round-2-private

Each sandbox has the following directory layout:

```
+-- common
|
+-- run_in
|   +--- sample
|   |    +-- input
|   |    |   +--- training
|   |    |   +--- testing
|   |    |
|   |    +-- output
|   |
|   +--- test
|
+-- eval_in
```

...where:

- "common" contains files whose data is invariant across all sandbox
instances

- "run_in/sample" is a dataset for which we have provided a train / test
split, and have furthermore provided the output of that train / test split
when used with both the reference CRF model and the baseline McAuley method.
(It may be that the demo set is too small for CRF or McAuley, in which case,
dummy values will be supplied.)

***
In general, we supply all truth labels for data in run_in/sample. We do not
supply the labels for the data in run_in/test.
***

- "run_in/sample/input/training" is the sample training data.

- "run_in/sample/input/testing" is the sample test data to be run against models
trained on run_in/sample/input/training.

- "run_in/sample/output" contains the results of reference CRF / baseline McAuley
on run_in/sample/testing when trained on run_in/sample/input/training.

- "run_in/test" is the dataset which will be processed after the teams have submitted
a model for evaluation, and is functionally equivalent to run_in/sample/input/testing.

- "eval_in" is where a CP6 solution writes its output for the data in "run_in/test",
and is functionally equivalent to run_in/sample/output.

The teams may use as much of the data as they wish in
run_in/sample/{training, testing} for preparing their model to be
evaluated against the data in run_in/test.

**Comment by collinsr** [comment]
I'm still perplexed. My impression is that the goal of the sandbox is to have the teams populate the eval_in directory, because this is what's evaluated. In turn, this requires that we decide what data goes in run_in/test. This data can either come from the released data or the sequestered data.

For the demo sandboxes, this makes sense: All data in the sandbox comes from the released data; we provide all the labels for a small set of images distributed over the sample/input/{training,testing} and the run_in/test set. They can score this locally and validate the entire test / train / evaluation cycle.

For the private sandboxes, this also makes sense: we provide all the training data, and plug all the sequestered data in run_in/test. Thus we evaluate them on the sequestered data.

Where I'm getting stuck is on the public sandbox: what's it for?  What's intended to go in run_in/test? Who scores it?  If they score it, then they must have the labels for the run_in/test data, and it's a demo sandbox, just using all the released data.  If we score it, then we have to leak the sequestered data from private to populate the test data. 

Either of these is feasible; I'm just unsure of the intention.
